### PR TITLE
Value injector for RESTEasy

### PR DESCRIPTION
**Pull request by yegorius**
What do you think?

### CODE REVIEW

**Comment by victornoel** [comment]
Oh great! I'm going to review this soon!

**Comment by victornoel** [comment]
You should already try to enable the tests that involve the annotations, it's easy, there is just some empty overriding test method to remove normally.

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	private ValueInjector getValueInjector(Class type, Annotation[] annotations) {
+		final Pac4JProfile profile;
+		if ((profile = FindAnnotation.findAnnotation(annotations, Pac4JProfile.class)) != null) {
+			if (type.equals(Optional.class)) {
+				return new Pac4JValueInjector(
+						(req, resp) -> new ProfileManager(new J2EContext(req, resp)).get(profile.readFromSession())
+				);
+			} else {
+				return new Pac4JValueInjector(
+						(req, resp) -> new ProfileManager(new J2EContext(req, resp)).get(profile.readFromSession()).orElse(null)
```
Comment (victornoel) [inline src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: In the jersey implementation, I preferred to throw an `new WebApplicationException(401)` instead of returning null.
The rational is that if you want to retrieve a profile that can be potentially not there (i.e. you are not authenticated), then you use the Optional.
If not (i.e., you are expecting to be authenticated, then the profile shouldn't be null, then we should't allow for such situation, especially because it introduces a security risks.

Actually one of the test relies on this behaviour ;)

See https://github.com/pac4j/jax-rs-pac4j/blob/master/src/main/java/org/pac4j/jax/rs/jersey/features/Pac4JValueFactoryProvider.java#L246

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	public static class Pac4JValueInjector implements ValueInjector {
+		private final BiFunction<HttpServletRequest, HttpServletResponse, Object> provider;
+
+		Pac4JValueInjector(BiFunction<HttpServletRequest, HttpServletResponse, Object> provider) {
+			this.provider = provider;
+		}
+
+		public Object inject(HttpRequest request, HttpResponse response) {
```
Comment (victornoel) [inline src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: Add an `@Override` annotation here.

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			return inject();
+		}
+
+		public Object inject() {
```
Comment (victornoel) [inline src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: Add an `@Override` annotation here.

File: src/main/java/org/pac4j/jax/rs/pac4j/JaxRsProfileManager.java [hunk 4]
```diff
@@ -131,6 +131,10 @@ public PrincipalImpl(LinkedHashMap<String, CommonProfile> profiles) {
             this.roles = Collections.unmodifiableSet(rs);
         }
 
+        public CommonProfile getProfile() {
```
Comment (victornoel) [inline src/main/java/org/pac4j/jax/rs/pac4j/JaxRsProfileManager.java]: This is not related to this PR, right? I would prefer you remove it then :)

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
+package org.pac4j.jax.rs.resteasy;
```
Comment (victornoel) [inline src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: Move this class into org.pac4j.jax.rs.resteasy.features, it would make more sense with the organization of the rest of this library :)

**Comment by victornoel** [comment]
There are also some formatting problems... some tabs crept up :)

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java (hunk 3 above)
Comment (yegorius) [reply src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: ✓

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java (hunk 2 above)
Comment (yegorius) [reply src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: ✓

File: src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java (hunk 5 above)
Comment (yegorius) [reply src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java]: ✓

File: src/main/java/org/pac4j/jax/rs/pac4j/JaxRsProfileManager.java (hunk 4 above)
Comment (yegorius) [reply src/main/java/org/pac4j/jax/rs/pac4j/JaxRsProfileManager.java]: ✓

**Comment by yegorius** [comment]
I am investigating issues with tests right now.

**Comment by yegorius** [comment]
The only test which fails for me is `testInject()`. And it only fails when it is being run with other tests. If I launch this single test, then it passes.
I get a weird error message from Undertow: `UT005014: Failed to parse HTTP request`.
Which happens because this is  exactly(!) what it receives on the wire (I have copied the complete byte buffer from the Undertow request parser):
```
User-Agent: Java/1.8.0_152-release
Host: localhost:24257
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: close
Cookie: JSESSIONID=5icpg-Zdu3SOLgo2QXDn_s2X_zzO00y5auNK9hqu
```
At this moment, I have no clue.
I have added a delay in the very beginning of `testInject()` for 5000ms and all tests pass now, but this is obviously not a solution, it's more of a hint for us.

**Comment by victornoel** [comment]
Thanks for the exploration.
I suppose it is undertow's closing that is not properly synchronized with the execution of tests or something like that… I say that because I've noticed harmless (until now...) undertow errors in the logs at the end of the tests before.
I will try to find the time to fix this next week-end, but if you feel like it, don't hesitate to continue exploring and maybe finding the solution ;)

**Comment by yegorius** [comment]
I have ignored the failing test so that the build is successful now. I have also updated the documentation.
Would you mind merging this PR in this state and moving the test problem to a separate GitHub issue?

**Comment by victornoel** [comment]
yes, ok, let's do that

**Comment by victornoel** [comment]
Thanks again @yegorius :)

We should do a release soon then! Maybe 2.1.0 @leleuj, what do you think?

**Comment by leleuj** [comment]
Yes, it makes sense. It'a noticeable update. Starting...

**Comment by leleuj** [comment]
@victornoel The version is still 2.0.3, I think it should be 2.1.0 as it's a backward compatible change (not a bug fix), isn't it?

**Comment by victornoel** [comment]
yes, 2.1.0 is a good number! (that's what I was saying in my comment above ; )

**Comment by leleuj** [comment]
Yes, right. I just performed the release, it will show up in the Maven central repository in a few hours...
